crates/accelerate/src/gate_direction.rs

```diff
@@ -142,8 +157,505 @@ where
+                    clbits: PyTuple::empty_bound(py).unbind(),
+                    params: packed_inst.params_view().iter().cloned().collect(),
+                    extra_attrs: packed_inst.extra_attrs.clone(),
+                    #[cfg(feature = "cache_pygates")]
+                    py_op: packed_inst.py_op.clone(),
+                },
+                sort_key: "".into_py(py),
+            },
+            DAGNode { node: None },
+        ),
+    )?;
+
+    dag.has_calibration_for(py, dag_op_node.borrow(py))
+}
+
+// Return a replacement DAG for the given standard gate in the supported list
+// TODO: optimize it by caching the DAGs of the non-parametric gates and caching and
+// mutating upon request the DAGs of the parametric gates
+fn replace_dag(
+    py: Python,
+    std_gate: StandardGate,
+    inst: &PackedInstruction,
+) -> PyResult<DAGCircuit> {
+    let replacement_dag = match std_gate {
+        StandardGate::CXGate => cx_replacement_dag(py),
+        StandardGate::ECRGate => ecr_replacement_dag(py),
+        StandardGate::CZGate => cz_replacement_dag(py),
+        StandardGate::SwapGate => swap_replacement_dag(py),
+        StandardGate::RXXGate => rxx_replacement_dag(py, inst.params_view()),
+        StandardGate::RYYGate => ryy_replacement_dag(py, inst.params_view()),
+        StandardGate::RZZGate => rzz_replacement_dag(py, inst.params_view()),
+        StandardGate::RZXGate => rzx_replacement_dag(py, inst.params_view()),
+        _ => panic!("Mismatch in supported gates assumption"),
+    };
+
+    replacement_dag
+}
+
+//###################################################
+// Utility functions to build the replacement dags
+// TODO: replace once we have fully Rust-friendly versions of QuantumRegister, DAGCircuit and ParemeterExpression
+#[inline]
+fn create_qreg(py: Python<'_>, size: u32) -> PyResult<Bound<'_, PyAny>> {
+    imports::QUANTUM_REGISTER.get_bound(py).call1((size,))
+}
+
+#[inline]
+fn qreg_bit<'py>(py: Python, qreg: &Bound<'py, PyAny>, index: u32) -> PyResult<Bound<'py, PyAny>> {
+    qreg.call_method1(intern!(py, "__getitem__"), (index,))
+}
+
+#[inline]
+fn std_gate(py: Python, gate: StandardGate) -> PyResult<Py<PyAny>> {
+    gate.create_py_op(
+        py,
+        None,
+        &ExtraInstructionAttributes::new(None, None, None, None),
+    )
+}
+
+#[inline]
+fn parameterized_std_gate(py: Python, gate: StandardGate, param: &[Param]) -> PyResult<Py<PyAny>> {
+    debug_assert!(param.len() == 1); //
+    gate.create_py_op(
+        py,
+        Some(param),
+        &ExtraInstructionAttributes::new(None, None, None, None),
+    )
+}
+
+#[inline]
+fn apply_op_back(
+    py: Python,
+    dag: &mut DAGCircuit,
+    op: &Py<PyAny>,
+    qargs: &Vec<&Bound<PyAny>>,
+) -> PyResult<()> {
+    dag.py_apply_operation_back(
+        py,
+        op.bind(py).clone(),
+        Some(TupleLikeArg::extract_bound(&PyTuple::new_bound(py, qargs))?),
+        None,
+        false,
+    )?;
+
+    Ok(())
+}
```

Comment:
Hmm... Could we perhaps use something like `DAGCircuit::apply_operation_back()` here? To take advantage of the rust native methods. Now I understand it makes it a little more annoying since we need to copy over the `Param`s and the `ExtraInstAttributes` but at least you remove the Python abstraction.